src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultFileChangeTracker.cs

```diff
@@ -0,0 +1,137 @@
+
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+        private readonly ErrorReporter _errorReporter;
+        private readonly IVsFileChangeEx _fileChangeService;
+        private uint _fileChangeCookie;
+
+        public override event EventHandler<FileChangeEventArgs> Changed;
+
+        public DefaultFileChangeTracker(
+            string filePath,
+            ForegroundDispatcher foregroundDispatcher,
+            ErrorReporter errorReporter,
+            IVsFileChangeEx fileChangeService)
+        {
+            if (string.IsNullOrEmpty(filePath))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(filePath));
+            }
+
+            if (foregroundDispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (errorReporter == null)
+            {
+                throw new ArgumentNullException(nameof(errorReporter));
+            }
+
+            if (fileChangeService == null)
+            {
+                throw new ArgumentNullException(nameof(fileChangeService));
+            }
+
+            FilePath = filePath;
+            _foregroundDispatcher = foregroundDispatcher;
+            _errorReporter = errorReporter;
+            _fileChangeService = fileChangeService;
+            _fileChangeCookie = VSConstants.VSCOOKIE_NIL;
+        }
+
+        public override string FilePath { get; }
+
+        public override void StartListening()
+        {
+            try
+            {
+                if (_fileChangeCookie == VSConstants.VSCOOKIE_NIL)
+                {
+                    var hr = _fileChangeService.AdviseFileChange(
+                        FilePath,
+                        FileChangeFlags,
+                        this,
+                        out _fileChangeCookie);
+
+                    Marshal.ThrowExceptionForHR(hr);
+                }
+            }
+            catch (Exception exception)
+            {
+                _errorReporter.ReportError(exception);
+            }
+        }
+
+        public override void StopListening()
+        {
+            try
+            {
+                if (_fileChangeCookie != VSConstants.VSCOOKIE_NIL)
+                {
+                    var hr = _fileChangeService.UnadviseFileChange(_fileChangeCookie);
+                    Marshal.ThrowExceptionForHR(hr);
+                    _fileChangeCookie = VSConstants.VSCOOKIE_NIL;
+                }
+            }
+            catch (Exception exception)
+            {
+                _errorReporter.ReportError(exception);
+            }
+        }
+
+        public int FilesChanged(uint fileCount, string[] filePaths, uint[] fileChangeFlags)
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            Debug.Assert(filePaths.Length == 1, "We only subscribed to 1 file change therefore we should only have 1 filepath.");
```

Comment:
Technically, it is not a guarantee that the length here will be 1. Looking at [this](http://index/?query=SVsFileChangeEx&rightProject=CloudTools(Repo)&file=Main.CAT%5Csrc%5CCommon%5CExternals%5CVSSDK.VS140%5CVisualStudioIntegration%5CTools%5CWizards%5CTemplates%5C1033%5CCS%5CEditorPane.cs&line=1844), I feel we should leave this as [before](https://github.com/aspnet/Razor/pull/1805/files#diff-d269dc5729aa0fa13d3ebb38462253efL155) (loop through all the filePaths). Although, I don't know if we want to go the extent of having a timer delay.

Reply:
Hmm, going to remove the length check then and just fire events for everything